src/plugins/merkleMemberships/client/index.ts

```diff
@@ -1,112 +1,173 @@
-          [ZkProgram.PublicInput, ZkProgram.TreeWitness],
-          Field
-        ]
-      ) => {
-        const privInput = new ZkProgram.PrivateInput({ witness, secret });
-        return acc.then(
-          O.match(
-            () => ZkProgram.Program.baseCase(root, privInput).then(O.some),
-            (prev) =>
-              ZkProgram.Program.inductiveCase(root, prev, privInput).then(
-                O.some
-              )
-          )
-        );
-      }
-    )(A.zip(publicInput, secretInput));
-
-    return O.match(
-      () => {
-        throw 'empty input list';
-      }, // TODO: make it pure
-      (p: ZkProof) => p.toJSON()
-    )(proof);
+    const computeRecursiveProof =
+      (l: Array<[[ZkProgram.PublicInput, ZkProgram.TreeWitness], Field]>) =>
+      (sp: ZkProof): TaskEither<string, ZkProof> =>
+        A.foldLeft(
+          // Pattern matching, not actually folding
+          () => TE.right(sp),
+          (
+            [[root, witness], secret]: [
+              [ZkProgram.PublicInput, ZkProgram.TreeWitness],
+              Field
+            ],
+            tail
+          ) =>
+            pipe(
+              fromFailablePromise(
+                () =>
+                  ZkProgram.Program.inductiveCase(
+                    root,
+                    sp,
+                    new ZkProgram.PrivateInput({ witness, secret })
+                  ),
+                'failed in inductive case'
+              ),
+              TE.chain((proof) => computeRecursiveProof(tail)(proof))
+            )
+        )(l);
+
+    const computeFinalProof = (
+      pl: NonEmptyArray<[ZkProgram.PublicInput, ZkProgram.TreeWitness]>,
+      sl: NonEmptyArray<Field>
+    ): TaskEither<string, ZkProof> => {
+      const l = NE.zip(pl, sl);
+      const h = NE.head(l);
+      const t = NE.tail(l);
+      return pipe(computeBaseProof(h), TE.chain(computeRecursiveProof(t)));
+    };
+
+    return pipe(
+      TE.Do,
+      TE.tap(() =>
+        publicInput.length != secretInput.length
+          ? TE.left('unmatched public/secret input list')
+          : TE.right(undefined)
+      ),
+      TE.bind('publicInputNE', () =>
+        TE.fromOption(() => 'public input list empty')(
+          NE.fromArray(publicInput)
+        )
+      ),
+      TE.bind('secretInputNE', () =>
+        TE.fromOption(() => 'private input list empty')(
+          NE.fromArray(secretInput)
+        )
+      ),
+      TE.chain(({ publicInputNE, secretInputNE }) =>
+        computeFinalProof(publicInputNE, secretInputNE)
+      ),
+      TE.map((finalProof) => finalProof.toJSON())
+    );
   }
 
-  // async fetchPublicInputs(
-  //   args: PublicInputArgs
-  // ): Promise<Array<[ZkProgram.PublicInput, ZkProgram.TreeWitness]>> {
-  //   const mkUrl = (treeRoot: Field, leafIndex: bigint) =>
-  //     `${this.cfg.baseUrl}/getWitness/${treeRoot
-  //       .toBigInt()
-  //       .toString()}/${leafIndex.toString()}`;
+  // fetchPublicInputs(
```

Comment:
Why this is left commented out?

Reply:
It's from the previously pr: https://github.com/mlabs-haskell/MinAuth/pull/2#discussion_r1352813447